### PR TITLE
s0ix-selftest-tool: init at 2022-11-04

### PR DESCRIPTION
**Pull request by adamcstephens**
###### Description of changes

Init a package for https://github.com/intel/S0ixSelftestTool

This is a tool for running a set of checks to validate Intel machines are correctly entering the lower power and/or s2idle states. Offering this package up as it requires a number of dependencies and does not work out of the box on NixOS.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  src = fetchFromGitHub {
+    owner = "intel";
+    repo = "S0ixSelftestTool";
+    rev = "1b6db3c3470a3a74b052cb728a544199661d18ec";
+    hash = "sha256-w97jfdppW8kC8K8XvBntmkfntIctXDQCWmvug+H1hKA=";
+  };
+
+  # don't use the bundled turbostat binary
+  patchPhase = ''
+    rm turbostat
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:33]: Looks like this rm is not needed?

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    substituteInPlace s0ix-selftest-tool.sh --replace '"$DIR"/turbostat' 'turbostat'
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:34]: Better not override entire phases, just do this in `postPatch`

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  '';
+
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
+    mkdir -p $out/bin
+    chmod +x s0ix-selftest-tool.sh
+    cp s0ix-selftest-tool.sh $out/bin/s0ix-selftest-tool
+  '';
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:44]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    install -Dm555 s0ix-selftest-tool.sh "$out/bin/s0ix-selftest-tool"
    
    runHook postInstall
  '';
```

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  solutions = {
+    default = {
+      scripts = ["bin/s0ix-selftest-tool"];
+      interpreter = "${lib.getExe bash}";
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:49]:
```suggestion
      interpreter = lib.getExe bash;
```

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        gnugrep
+        gnused
+        linuxPackages.turbostat
+        pciutils
+        powertop
+        util-linux
+        xorg.xset
+        xxd
+      ];
+      fake.external = ["dmesg" "sudo"];
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:64]: `dmesg` is not a suid wrapper on NixOS.

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      scripts = ["bin/s0ix-selftest-tool"];
+      interpreter = "${lib.getExe bash}";
+      inputs = [
+        acpica-tools
+        bc
+        coreutils
+        gawk
+        gnugrep
+        gnused
+        linuxPackages.turbostat
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:57]: Or shall this package also be under `linuxPackages`, in order to use the correct version of turbostat for all kernel version.

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        "cannot:${powertop}/bin/powertop"
+        "cannot:${util-linux}/bin/rtcwake"
+        "cannot:${linuxPackages.turbostat}/bin/turbostat"
+      ];
+    };
+  };
+
+  meta = with lib; {
+    homepage = "https://github.com/intel/S0ixSelftestTool";
+    description = "A tool for testing the S2idle path CPU Package C-state and S0ix failures in Linux OS for Intel® Client platforms, it also supports the basic runtime PC10 status check. ";
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:75]: A bit too wordy, also no trailing dot.

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -39006,4 +39006,6 @@ with pkgs;
   resgate = callPackage ../servers/resgate { };
 
   oversteer = callPackage ../applications/misc/oversteer { };
+
+  s0ix-selftest-tool = callPackage ../tools/system/s0ix-selftest-tool { };
```
Comment (NickCao) [inline pkgs/top-level/all-packages.nix:39010]: Keep the file roughly alphabetically sorted, don't just add it at bottom.

Comment (adamcstephens) [reply pkgs/top-level/all-packages.nix:39010]: There is little order in this file, and there are probably two dozen places that would be "roughly alphabetical". I moved it up to a different area anyway.

File: pkgs/tools/system/s0ix-selftest-tool/default.nix (hunk 6 above)
Comment (adamcstephens) [reply pkgs/tools/system/s0ix-selftest-tool/default.nix:56]: I considered following that pattern, and can still move it there if necessary. Upstream is bundling their own turbostat which is being used across all kernel versions, which implies to me the specific kernel version doesn't matter.

**Comment by adamcstephens** [comment]
Thanks for your review @NickCao. I addressed all the feedback, minus the question of whether we want to put this in a kernel-specific package.

**Comment by NickCao** [comment]
Let me check about that tomorrow, really need some rest.

File: pkgs/tools/system/s0ix-selftest-tool/default.nix (hunk 6 above)
Comment (NickCao) [reply pkgs/tools/system/s0ix-selftest-tool/default.nix:56]: There appears to be little (not tied to kernel version) references to linuxPackages in nixpkgs. But I do find one in `pkgs/desktops/gnome/misc/gnome-applets/default.nix`, I suppose this is acceptable as long as it works.

File: pkgs/tools/system/s0ix-selftest-tool/default.nix [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        gnugrep
+        gnused
+        linuxPackages.turbostat
+        pciutils
+        powertop
+        util-linux
+        xorg.xset
+        xxd
+      ];
+      fake.external = ["sudo"];
```
Comment (NickCao) [inline pkgs/tools/system/s0ix-selftest-tool/default.nix:63]: Found this line in the resulting script.
`pciests1=$(sudo xxd -ps -l2 -s 0x32a "${PCI_SYSFS}"/"${dev}"/config)`
Apparently xxd is not correctly resholved.
As the script must be run as root, what about just remove that sudo call.

Comment (adamcstephens) [reply pkgs/tools/system/s0ix-selftest-tool/default.nix]: Removed the sudo call and xxd resholved now.

File: pkgs/tools/system/s0ix-selftest-tool/default.nix (hunk 6 above)
Comment (adamcstephens) [reply pkgs/tools/system/s0ix-selftest-tool/default.nix:58]: Yeah, not much prior art on this. It does build and work as-is in my testing, so I would vote for leaving it standalone rather than putting under every kernel version.

**Review by NickCao — APPROVED** [review]
